src/rgw/rgw_bucket.cc

```diff
@@ -2530,14 +2531,37 @@ int RGWBucketInstanceMetadataHandler::do_put(RGWSI_MetaBackend_Handler::Op *op,
   return do_put_operate(&put_op);
 }
 
-void init_default_bucket_layout(CephContext *cct, RGWBucketInfo& info, const RGWZone& zone) {
-  info.layout.current_index.gen = 0;
-  info.layout.current_index.layout.normal.hash_type = rgw::BucketHashType::Mod;
-  info.layout.current_index.layout.type = rgw::BucketIndexType::Normal;
+void init_default_bucket_layout(CephContext *cct, rgw::BucketLayout& layout,
+				const RGWZone& zone,
+				std::optional<uint32_t> shards,
+				std::optional<rgw::BucketIndexType> type) {
+  layout.current_index.gen = 0;
+  layout.current_index.layout.normal.hash_type = rgw::BucketHashType::Mod;
 
-  info.layout.current_index.layout.normal.num_shards = (
-      cct->_conf->rgw_override_bucket_index_max_shards > 0 ?
-      cct->_conf->rgw_override_bucket_index_max_shards : zone.bucket_index_max_shards);
+  layout.current_index.layout.type =
+    type ? *type : rgw::BucketIndexType::Normal;
+
+  if (type) {
+    layout.current_index.layout.type = *type;
+  } else {
+    layout.current_index.layout.type = rgw::BucketIndexType::Normal;
+  }
```

Comment:
oops, this initialization is duplicated in two different forms

optional::value_or(default) looks like a good fit here:
```c++
layout.current_index.layout.type = type.value_or(rgw::BucketIndexType::Normal);
```